### ISSUE TITLE
GraphQL Template format error

### ISSUE THREAD

**Issue by wbroek** [aws-amplify/amplify-cli#1526]
**Describe the bug**
When updating my GraphQL schema (added one boolean field) I get the error
`UPDATE_FAILED      apimcggraphql AWS::CloudFormation::Stack Mon May 27 2019 15:59:33 GMT+0200 (Central European Summer Time) Template format error: Output GetAttCarepathItemArticleTableStreamArn is malformed.  The Name field of Export must not depend on any resources, imported values, or Fn::GetAZs.`

The `CarepathItemArticle` is in my model but was there before and no changes made for that part of the model

***Model***
_(not the whole model which is larger)_
```
type CarepathItem @model
    @auth(rules: [{ allow: groups, groups: ["admin"], operations: [create, update, delete]}])
{
  id: ID!
  message: String
  notification: Boolean
  repeat: Boolean
  #...#
  articles: [CarepathItemArticle] @connection(name: "CarepathsArticles")
}

type CarepathItemArticle @model(queries: null) 
{
  id: ID!
  path: CarepathItem! @connection(name: "CarepathsArticles")
  article: Article! @connection(name: "ArticlesCarepaths")
}

# Article
type Article @model 
{
  id: ID!
  title: String!
  video: String
  content: String
  picture: S3Object
  createdAt: AWSDateTime
  paths: [CarepathItemArticle] @connection(name: "ArticlesCarepaths")
}
```

**Desktop Cli:**
 - OS: MacOSX 10.14.5
 - Version CLI  1.16.11

**Additional context**
What was interesting is that when I looked at the generated cloud formation (in `backend/api/mcggraphl/build/cloudformation-template.yml`) he generated two new items which where not there before? And if you remove those two keys (GetAttCarepathItemArticleTableStreamArn and GetAttCarepathItemArticleTableName) it seems to work. Where are those generated?

```
"Outputs": {
        "GraphQLAPIIdOutput": {
            "Description": "Your GraphQL API ID.",
            "Value": {
                "Fn::GetAtt": [
                    "GraphQLAPI",
                    "ApiId"
                ]
            },
            "Export": {
                "Name": {
                    "Fn::Join": [
                        ":",
                        [
                            {
                                "Ref": "AWS::StackName"
                            },
                            "GraphQLApiId"
                        ]
                    ]
                }
            }
        },
        "GraphQLAPIEndpointOutput": {
            "Description": "Your GraphQL API endpoint.",
            "Value": {
                "Fn::GetAtt": [
                    "GraphQLAPI",
                    "GraphQLUrl"
                ]
            },
            "Export": {
                "Name": {
                    "Fn::Join": [
                        ":",
                        [
                            {
                                "Ref": "AWS::StackName"
                            },
                            "GraphQLApiEndpoint"
                        ]
                    ]
                }
            }
        },
        "GetAttCarepathItemArticleTableStreamArn": {
            "Description": "Your DynamoDB table StreamArn.",
            "Value": {
                "Fn::GetAtt": [
                    "CarepathItemArticleTable",
                    "StreamArn"
                ]
            },
            "Export": {
                "Name": {
                    "Fn::Join": [
                        ":",
                        [
                            {
                                "Ref": "AppSyncApiId"
                            },
                            "GetAtt",
                            "CarepathItemArticleTable",
                            "StreamArn"
                        ]
                    ]
                }
            }
        },
        "GetAttCarepathItemArticleTableName": {
            "Description": "Your DynamoDB table name.",
            "Value": {
                "Ref": "CarepathItemArticleTable"
            },
            "Export": {
                "Name": {
                    "Fn::Join": [
                        ":",
                        [
                            {
                                "Ref": "AppSyncApiId"
                            },
                            "GetAtt",
                            "CarepathItemArticleTable",
                            "Name"
                        ]
                    ]
                }
            }
        },
        "GraphQLAPIKeyOutput": {
            "Description": "Your GraphQL API key. Provide via 'x-api-key' header.",
            "Value": {
                "Fn::GetAtt": [
                    "GraphQLAPIKey",
                    "ApiKey"
                ]
            },
            "Export": {
                "Name": {
                    "Fn::Join": [
                        ":",
                        [
                            {
                                "Ref": "AWS::StackName"
                            },
                            "GraphQLApiKey"
                        ]
                    ]
                }
            },
            "Condition": "ShouldCreateAPIKey"
        }
    },
```

_btw great work you all on the CLI_

**Comment by mikeparisstuff** [comment]
@wbroek Hmm, those outputs should be in the same stack as your model. There was a change recently by @kstro21 that always adds these outputs to fix a bug related to missing ARNs when trying to remove the `@searchable` directive.

Was this project migrated from the earlier version of the CLI? There would be a `transform.conf.js` file that specifies migration information if so. The relevant code is here https://github.com/aws-amplify/amplify-cli/blob/31d4a89173a1cc068160c13cdaaa68f4b7e4f64f/packages/graphql-dynamodb-transformer/src/DynamoDBModelTransformer.ts#L145-L157. 

Thanks for the support. Let's get you unblocked.

**Comment by wbroek** [comment]
@mikeparisstuff thanks for the response.
Yes the project was migrated (using Amplify since version 0.1.13) from an old version; contents of `transform.conf.js`
```
{
    "Migration": {
        "V1": {
            "Resources": [
                "UserTable",
                "CarepathTable",
                "CarepathItemArticleTable",
                "ArticleTable",
                "ArticleCategorieTable",
                "ArticleCategoriesTable",
                "QuestionnaireTable",
                "QuestionnaireAnswerTable"
            ]
        }
    }
}
```
Can also confirm that if i remove the `transform.conf.js` (or only the CarepathItemArticleTable item) it does not generate the faulty output in Ouputs.

Can we safely remove the transform.conf.js`??

**Comment by kstro21** [comment]
Hi @wbroek, if you remove **transform.conf.js**, your DynamoDB tables, Elasticsearch instance, and other resources, will be moved from the template where are defined now, meaning all will be dropped and created from 0. You might be losing data in the process.

**Comment by wbroek** [comment]
@kstro21 thanks for the reply. Question remains how to solve it then? Without losing data of course
Hope hearing from you

**Comment by mikeparisstuff** [comment]
@wbroek Thanks for the details. I will take a look and see how to unblock you.

**Comment by mikeparisstuff** [comment]
@wbroek You should be able to fix your project by changing the contents of `transform.conf.json` to be:

```json
{
    "Migration": {
        "V1": {
            "Resources": [
                "^UserTable$",
                "^CarepathTable$",
                "^CarepathItemArticleTable$",
                "^ArticleTable$",
                "^ArticleCategorieTable$",
                "^ArticleCategoriesTable$",
                "^QuestionnaireTable$",
                "^QuestionnaireAnswerTable$"
            ]
        }
    }
}
```

The names under **Migration.V1.Resources** are used to specify which resources should be hoisted up to the top level in the event that they were created before nested stacks support released in the CLI. Adding the `^` and `$` characters to the beginning and end of the strings respectively will stop the transform from thinking that the "GetAttCarepathItemArticleTableStreamArn" should be pulled up into the root stack. 

I'd recommend testing this by making the change to `transform.conf.json`, running `amplify api gql-compile`, and then check to make sure the `build/cloudformation-template.json` still contains all the `AWS::DynamoDB::Table` resources as expected. You should also no longer see `GetAttCarepathItemArticleTableStreamArn` in `build/cloudformation-template.json`.

The transform will not overwrite the transform.conf.json file so this change should permanently fix the problem for you. Let me know if this unblocks you and I'll make sure this is recorded so that we can fix any related code to prevent this in the future.

**Comment by wbroek** [comment]
@mikeparisstuff thank you (the removed comment did not work? Here it didn't indeed)

**Comment by kstro21** [comment]
@mikeparisstuff were you able to reproduce the issue described by @wbroek? I would like to submit a PR to solve it but I'm having a hard time. Probably is time to implement what you suggested once here https://github.com/aws-amplify/amplify-cli/pull/1316#issuecomment-486875664

**Comment by kstro21** [comment]
Ok, I was able to reproduce it. I had to use the migrations shared by @wbroek. Now looking for the fix.

**Comment by wbroek** [comment]
@kstro21 any news, still got it (workaround I use is removing the wrong keys in the cloudformation-template) with version 1.7.3

**Comment by kstro21** [comment]
Hey, @wbroek, sorry for the late, I've been short of time. 

No success yet finding the solution. Your `type CarepathItemArticle` is generating a `CarepathItemArticleTable` resource mapped to a **stack** that is not defined. That is the reason the **outputs** ends up in the **root stack**.

I've managed to **always** map the **outputs** to the corresponding **stack** as suggested by @mikeparisstuff in https://github.com/aws-amplify/amplify-cli/pull/1316#issuecomment-486875664 but then it throws an error in line 126 https://github.com/aws-amplify/amplify-cli/blob/38d0ce85b4dcae0ee591292730ac8e93e448c9b7/packages/graphql-transformer-core/src/util/splitStack.ts#L123-L127

The error says
>TypeError: Cannot read property 'Outputs' of undefined

I will need some more time to debug, probably another week as I'm short of time, and probably some assistance from @mikeparisstuff on the desition making for the solution.

**Comment by wbroek** [comment]
This looks like to be fixed in 1.7.8 @kstro21, the output is now correct and the GetAttCarepathItemArticleTableStreamArn is not added anymore to the output key and the error is no longer there.
BUT now I got the same as mentioned in https://github.com/aws-amplify/amplify-cli/issues/1581 ; the CarepathItemArticle is added differently and thus getting the error 
`CREATE_FAILED      um-20190103001419-apiumcggraphql-YVII6RRKP3B8-CarepathItemArticle-11IX5XA8NKSB8 AWS::CloudFormation::Stack Wed Jul 10 2019 12:26:18 GMT+0200 (Central European Summer Time) Export with name ppu3w6g6f5b4lpnp2cgmeuz7ie:GetAtt:CarepathItemArticleDataSource:Name is already exported by stack um-20190103001419-apiumcggraphql-YVII6RRKP3B8-Article-U8IOLMJC8KXK`
Will close this one and add my finding to #1581
